specification/README.md

```diff
@@ -16,5 +16,62 @@ The list of Ably Asset Tracking specific error codes.
+The [locations](/test-resources/geo/schemas/location.json) are sent in the [GeoJSON](https://geojson.org/) format. 
+
+## Core architecture decisions
+
+### Reusing stopped SDK instances
+
+After the SDK instance is stopped it can never be used again. 
+If customers want to use the SDK again they will have to create a new instance of it.
+
+### Securing SDK from asynchronous access
+
+#### General approach
+
+Because mobile development is asynchronous in nature we had to take special measures to ensure that our SDK is safe to be used from multiple threads.
+Instead of choosing a mutex approach, we chose the approach of a synchronous events queue (similar to how the JS engine is implemented) to secure the SDK.
+Therefore, we should always aim to use the queue to perform operations that require synchronization and manipulate the SDK state.
+
+#### How the queue works
+
+When you have to perform work in the SDK you should enqueue the work in the synchronous queue.
+Any work queued will be taken out from the queue in the FIFO order and processed one by one.
+Additionally, the state of the SDK is only accessible from within the queue.
+This ensures that the state is always manipulated in a secure synchronous manner.
+
+#### Running asynchronous operations from the queue
+
+Sometimes the work we perform on the queue will require to run asynchronous operations.
+As a general rule we do not want to block queue and wait for async operations to complete as we can be doing other work during that time.
+When the async operation completes we should resume the work that triggered it by queueing new work.
+
+#### When SDK state can be manipulated
+
+The SDK state should be only visible from within the queue (just like a local variable of a function).
+The SDK state should only be accessed and modified from within the safe synchronous queue thread. 
+This will ensure there are no races and the state is always synchronized.
+You should never manipulate the state from the asynchronous operation listeners, ideally the queue should be implemented in a way that will prevent it by design.
+If you need to access or modify the state after some async operation completes, you should queue new work to the queue and perform the state manipulation from it.
+
+#### Stopping the queue
+
+The only exception to the rule that asynchronous operations shouldn't block the queue is when the queue is being stopped.
+If we wanted to wait for async operations after the queue stop operation was started we would have to introduce some kind of "stopping" state.
+In this state the queue would not be stopped yet but it won't process any work as well since it is being stopped.
+Instead when the queue is being stopped, the work is allowed to block the queue and wait for any async operations required to properly stop the SDK.
```

Comment:
I _think_ that when you say "the work" here you mean the work of stopping the publisher. Is that right? Might there be a way to make it a bit clearer?

Reply:
Good point 😉 https://github.com/ably/ably-asset-tracking-common/pull/70/commits/f2b3db41ca5208e1be1d08a71faf4a4175ad44e4